contrib/opencensus-ext-ocagent/opencensus/ext/ocagent/stats_exporter/__init__.py

```diff
@@ -0,0 +1,246 @@
+        metric_descriptor.MetricDescriptorType.CUMULATIVE_DOUBLE:
+        metrics_pb2.MetricDescriptor.CUMULATIVE_DOUBLE,
+        metric_descriptor.MetricDescriptorType.CUMULATIVE_DISTRIBUTION:
+        metrics_pb2.MetricDescriptor.CUMULATIVE_DISTRIBUTION,
+        metric_descriptor.MetricDescriptorType.GAUGE_INT64:
+        metrics_pb2.MetricDescriptor.GAUGE_INT64,
+        metric_descriptor.MetricDescriptorType.GAUGE_DOUBLE:
+        metrics_pb2.MetricDescriptor.GAUGE_DOUBLE,
+        metric_descriptor.MetricDescriptorType.GAUGE_DISTRIBUTION:
+        metrics_pb2.MetricDescriptor.GAUGE_DISTRIBUTION,
+        metric_descriptor.MetricDescriptorType.SUMMARY:
+        metrics_pb2.MetricDescriptor.SUMMARY,
+    }.get(descriptor_type, metrics_pb2.MetricDescriptor.UNSPECIFIED)
+
+
+class ExportRpcHandler(object):
+    """Manages the rpc to the exporter service.
+
+    :type client: class:`~.metrics_service_pb2_grpc.MetricsServiceStub`
+    :param client: metrics export client
+
+    :type service_name: str
+    :param service_name: name of the service
+
+    :type host_name: str
+    :param host_name: name of the host (machine or host name)
+    """
+
+    def __init__(self, client, service_name, host_name=None):
+        self._initialized = False
+        self._initial_request = None
+        self._rpc = bidi.BidiRpc(client.Export, lambda: self._initial_request)
+        self._node = utils.get_node(service_name, host_name)
+
+    def send(self, request):
+        """Dispatches incoming request on rpc.
+
+        Initializes rpc if necessary and dispatches incoming request.  If a rpc
+        error is thrown, this function will attempt to recreate the stream and
+        retry sending given request once.
+
+        :type request: class:
+          `~.metrics_service_pb2.ExportMetricsServiceRequest`
+        :param request: incoming export request
+        """
+        if not self._initialized:
+            self._initialize(request)
+            return
+
+        try:
+            self._rpc.send(request)
+        except grpc.RpcError as e:
+            logging.info('Found rpc error %s', e, exc_info=True)
+            # If stream has closed due to error, attempt to reopen with the
+            # incoming request.
+            self._initialize(request)
+
+    def _initialize(self, request):
+        """Initializes the exporter rpc stream."""
+
+        # Add node information on the first request dispatched on a stream.
+        request.node.MergeFrom(self._node)
+        self._initial_request = request
+
+        self._rpc.open()
+        self._initialized = True
+
+
+def _create_stub(endpoint):
+    return metrics_service_pb2_grpc.MetricsServiceStub(
+        grpc.insecure_channel(endpoint))
+
+
+def new_stats_exporter(service_name,
+                       hostname=None,
+                       endpoint=None,
+                       interval=None):
+    """Create a ocagent exporter and attaches it to the exporter thread.
```

Comment:
s/attaches/attach, but also note that this creates a new exporter thread now.

Reply:
Updated to note thread creation.